File: Containers/base-r/ai4e_api_tools/task_management/api_task.R
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
library(httr)

AddTask<-function(req){
  taskId = 1

  newTask <- data.frame(
    TaskId = taskId,
    Status = c("queued"),
    Timestamp = c(format(Sys.time(), "%b %d %X %Y")),
    Endpoint = "uri"
  )

  if (file.exists("tasks.csv")) {
    print("file exists")
    tasks<-read.csv("tasks.csv", stringsAsFactors=FALSE)

    lastId <- tail(tasks, 1)[,1]
    taskId = lastId + 1
    newTask$TaskId <- taskId
    
    write.csv(rbind(tasks, newTask), "tasks.csv", row.names=FALSE)
  }
  else {
    write.csv(newTask, "tasks.csv", row.names=FALSE)
  }

  return(newTask)
}

UpdateTaskStatus<-function(taskId, status){
  taskId = as.numeric(taskId)
  print(paste0("updating ", taskId, " to ", status))

  if (file.exists("tasks.csv")) {
    tasks<-read.csv("tasks.csv", stringsAsFactors=FALSE)

    tasks[taskId, 2] <- c(status)
    tasks[taskId, 3] <- c(format(Sys.time(), "%b %d %X %Y"))

    write.csv(tasks, "tasks.csv", row.names=FALSE)
  }
  else {
    newTask <- data.frame(
      TaskId = taskId,
      Status = c(status),
      Timestamp = c(format(Sys.time(), "%b %d %X %Y")),
      Endpoint = "uri"
    )

    write.csv(newTask, "tasks.csv", row.names=FALSE)
  }
}

CompleteTask<-function(taskId, status){
  UpdateTaskStatus(taskId, status)
}

FailTask<-function(taskId, status){
  UpdateTaskStatus(taskId, status)
}

AddPipelineTask<-function(taskId, organization_moniker, version, api_name, body) {
    next_url <- paste(version, organization_moniker, api_name, sep = "/")

    host <- Sys.getenv("LOCAL_NEXT_API_HOST_IN_PIPELINE")

    if (!is.empty(host)) {
      next_url <- paste(host, next_url, sep = "/")
    }

    tryCatch({
      res <- POST(next_url, body=body)

      if (status_code(r) != 200) {
        UpdateTaskStatus(taskId, paste0("Pipelining is not supported in a single node deployment, but the next service is: ", next_url))
        return(paste0("Pipelining is not supported in a single node deployment, but the next service is: ", next_url))
      }
    }, error = function(err) {
        print(err)
        UpdateTaskStatus(taskId, paste0("Pipelining is not supported in a single node deployment, but the next service is: ", next_url))
        return(paste0("Pipelining is not supported in a single node deployment, but the next service is: ", next_url))
    })

    return(res)
}

#* Get status of task by id
#* @param taskId The id of the task
#* @get /task/<taskId>
GetTaskStatus<-function(taskId){
  if (!file.exists("tasks.csv")) {
    return("Task not found.")
  }

  taskId = as.numeric(taskId)
  tasks<-read.csv("tasks.csv")
  tasks[taskId, 1:3]
}

# Please have an empty last line in the end; otherwise, you will see an error when starting a webserver

File: Containers/base-py/ai4e_api_tools/task_management/api_task.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
from datetime import datetime
import json
import os
import multiprocessing
from typing import Any, Dict
import uuid

import requests


print("Creating task manager.")

LOCAL_BLOB_TEST_DIRECTORY = os.getenv('LOCAL_BLOB_TEST_DIRECTORY', '.')

class TaskManager:

    # use a lock whenever we access task_status.json
    task_status_json_lock = multiprocessing.Lock()

    def __init__(self):
        self.status_dict = {}
        self.task_status_json_path = os.path.join(
            LOCAL_BLOB_TEST_DIRECTORY, 'task_status.json')

    def GetTaskId(self) -> str:
        return str(uuid.uuid4())

    def AddTask(self, request):
        status = {
            'TaskId': self.GetTaskId(),
            'Status': 'created',
            'Timestamp': datetime.strftime(
                datetime.utcnow(), "%Y-%m-%d %H:%M:%S"),
            'Endpoint': request.path
        }

        with self.task_status_json_lock:
            statuses = []
            if os.path.isfile(self.task_status_json_path):
                with open(self.task_status_json_path, 'r') as f:
                    statuses = json.load(f)
            statuses.append(status)

            with open(self.task_status_json_path, 'w') as f:
                json.dump(statuses, f)
        return status

    def UpdateTaskStatus(self, taskId: str, status: Any) -> None:
        with self.task_status_json_lock:
            if not os.path.isfile(self.task_status_json_path):
                raise ValueError('taskId "{}" is not found. Decorate your endpoint with an ai4e_service decorator or call AddTask(request) before UpdateTaskStatus.'.format(taskId))

            statuses = []
            with open(self.task_status_json_path, 'r') as f:
                statuses = json.load(f)

            task_status = None
            for rec_status in statuses:
                if rec_status['TaskId'] == taskId:
                    task_status = rec_status
                    break

            if task_status is None:
                raise ValueError('taskId "{}" is not found. Decorate your endpoint with an ai4e_service decorator or call AddTask(request) before UpdateTaskStatus.'.format(taskId))
            else:
                task_status['Status'] = status
                task_status['Timestamp'] = datetime.strftime(
                    datetime.utcnow(), "%Y-%m-%d %H:%M:%S")

            with open(self.task_status_json_path, 'w') as f:
                json.dump(statuses, f)

    def AddPipelineTask(self, taskId, organization_moniker, version, api_name, body):
        next_url = version + '/' + organization_moniker + '/' + api_name

        host = os.getenv('LOCAL_NEXT_API_HOST_IN_PIPELINE', '')

        if len(host) > 0:
            next_url = str(host) + '/' + str(next_url)

        r = requests.post(next_url, data=body)

        if r.status_code != 200:
            ai4e_service.api_task_manager.UpdateTaskStatus(taskId, "Pipelining is not supported in a single node deployment, but the next service is: " + next_url)
            return "Pipelining is not supported in a single node deployment, but the next service is: " + next_url
        else:
            return r.status_code

    def CompleteTask(self, taskId, status):
        self.UpdateTaskStatus(taskId, status)

    def FailTask(self, taskId, status):
        self.UpdateTaskStatus(taskId, status)

    def GetTaskStatus(self, taskId: str) -> Dict[str, Any]:
        with self.task_status_json_lock:
            if os.path.isfile(self.task_status_json_path):
                with open(self.task_status_json_path, 'r') as f:
                    statuses = json.load(f)

                for rec_status in statuses:
                    if rec_status['TaskId'] == taskId:
                        return rec_status

        status = {
            'TaskId': taskId,
            'Status': 'Not found.',
            'Timestamp': datetime.strftime(datetime.now(), "%Y-%m-%d %H:%M:%S"),
            'Endpoint': ''
        }
        return status
